eclipse-che/che-server
# Avoid caching when getting a raw file from raw.githubusercontent.com

Signed-off-by: Vitaliy Gulyy <vgulyy@redhat.com>



### What does this PR do?

At the moment it's difficult to test the flow on custom git servers, so the functionality is temporary enabled only for github.com.

To prevent caching of raw file content, when addressing the file via https://raw.githubusercontent.com it needs to use SHA of the last commit instead of the branch name.
It gives us the unique URI on each file change.

E.g., use
```
https://raw.githubusercontent.com/vitaliy-guliy/web-nodejs-sample/ce7b7ab0aa069e0bda52eac761e2d763ae78bd71/devfile.yaml
``` 
instead of
```
https://raw.githubusercontent.com/vitaliy-guliy/web-nodejs-sample/devfilev2/devfile.yaml
```

---
When client asks the content of `devfile.yaml` in `devfilev2` branch, che-server

- gets the last commit for `devfilev2` branch using GitHub rest API
```
https://api.github.com/repos/che-samples/web-nodejs-sample/commits?sha=devfilev2&page=1&per_page=1
```

- forms new URI to the raw file, where the branch name is replaced on the commit SHA
```
https://raw.githubusercontent.com/vitaliy-guliy/web-nodejs-sample/ce7b7ab0aa069e0bda52eac761e2d763ae78bd71/devfile.yaml
```

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

https://github.com/eclipse/che/issues/21184

### How to test this PR?

 
1. Deploy Che with
```
chectl server:deploy -p=minikube -i=quay.io/vgulyy/che-server:test-fetching-git-files
```
2. Create a workspace from git repository
3. Change the devfile, push changes
4. Within a minute, delete workspace and create a new one using the same factory link
5. Check if the changes have been applied

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.
